File: src/jsMain/kotlin/ast/ast.kt
@file:JsModule("ocdot-parser/lib/ast")
@file:JsNonModule
@file:Suppress(
    "INTERFACE_WITH_SUPERCLASS",
    "OVERRIDING_FINAL_MEMBER",
    "RETURN_TYPE_MISMATCH_ON_OVERRIDE",
    "CONFLICTING_OVERLOADS"
)
@file:JsQualifier("AST") // because AST is exported as namespace
package ast

import ClassParts
import Kind
import Kind1
import Readonly
import kotlin.js.*
import org.khronos.webgl.*
import org.w3c.dom.*
import org.w3c.dom.events.*
import org.w3c.dom.parsing.*
import org.w3c.dom.svg.*
import org.w3c.dom.url.*
import org.w3c.fetch.*
import org.w3c.files.*
import org.w3c.notifications.*
import org.w3c.performance.*
import org.w3c.workers.*
import org.w3c.xhr.*

external interface FilePosition {
    var offset: Number
    var line: Number
    var column: Number
}

external interface FileRange {
    var start: FilePosition
    var end: FilePosition
    var source: String
}


external interface LiteralExpectation {
    var type: String /* "literal" */
    var text: String
    var ignoreCase: Boolean
}


external interface ClassExpectation {
    var type: String /* "class" */
    var parts: ClassParts
    var inverted: Boolean
    var ignoreCase: Boolean
}

external interface AnyExpectation {
    var type: String /* "any" */
}

external interface EndExpectation {
    var type: String /* "end" */
}

external interface OtherExpectation {
    var type: String /* "other" */
    var description: String
}

external interface `T$0` {
    var OcDot: String /* "ocdot" */
    var Comment: String /* "comment" */
    var Ocnet: String /* "ocnet" */
    var Attribute: String /* "attribute" */
    var Attributes: String /* "attributes" */
    var Edge: String /* "edge" */
    var Node: String /* "node" */
    var NodeRef: String /* "node_ref" */
    var Subgraph: String /* "subgraph" */
    var EdgeSubgraph: String /* "edge_subgraph" */
    var Literal: String /* "literal" */
    var ClusterStatements: String /* "cluster_statements" */
    var TypeDefinitions: String /* "type_definitions" */
}


external interface `T$1` {
    var Places: String /* "places" */
    var Transitions: String /* "transitions" */
}

external var SubgraphSpecialTypes: `T$1`

external interface EdgeSubgraph : ASTBaseParent<dynamic /* Attribute | Attributes | Edge | Node | Subgraph | Comment */> {
    var id: Literal__0?
        get() = definedExternally
        set(value) = definedExternally
}

external interface `T$2` {
    var Normal : String  /* "->" */
    var Variable : String  /* "=>" */
}

external var OpTypes : `T$2`

external interface ParseOptions {
    var filename: String?
        get() = definedExternally
        set(value) = definedExternally
    var startRule: String?
        get() = definedExternally
        set(value) = definedExternally
    var tracer: Any?
        get() = definedExternally
        set(value) = definedExternally
    @nativeGetter
    operator fun get(key: String): Any?
    @nativeSetter
    operator fun set(key: String, value: Any)
}


external var Types: `T$0`

external fun isASTBaseNode(value: Any): Boolean

external interface ASTBaseNode {
    var type: String
    var location: FileRange
}

external interface ASTBaseParent<STMT : ASTBaseNode> : ASTBaseNode {
    var body: Array<STMT>
}

external interface Literal<T : String?> : ASTBaseNode {
    var value: T
    var quoted: dynamic /* Boolean | "html" */
        get() = definedExternally
        set(value) = definedExternally
}

external interface Literal__0 : Literal<String>

external interface OcDot : ASTBaseParent<dynamic /* OcNet | Comment */>

external interface OcNet : ASTBaseParent<dynamic /* Attribute | Attributes | Edge | Node | Subgraph | Comment */> {
    var id: Literal__0?
        get() = definedExternally
        set(value) = definedExternally
}

external interface KeyValue {
    var key: Literal__0
    var value: Literal__0
}

external interface Attribute : ASTBaseNode, KeyValue

@Suppress("NESTED_CLASS_IN_EXTERNAL_INTERFACE")
external interface Comment : ASTBaseNode {
    var kind: Kind
    var value: String

    interface `T$1` {
        var Block: String /* "block" */
        var Slash: String /* "slash" */
        var Macro: String /* "macro" */
    }

    companion object {
        var Kind: Readonly<`T$1`>
    }
}

@Suppress("NESTED_CLASS_IN_EXTERNAL_INTERFACE")
external interface Attributes : ASTBaseParent<dynamic /* Attribute | Comment */> {
    var kind: Kind1

    interface `T$2` {
        var Ocnet: String /* "ocnet" */
        var Edge: String /* "edge" */
        var Node: String /* "node" */
    }

    companion object {
        var Kind: Readonly<`T$2`>
    }
}

external interface NodeRef : ASTBaseNode {
    var id: Literal__0
    var port: Literal__0?
        get() = definedExternally
        set(value) = definedExternally
    var compass: Literal<String? /* "n" | "ne" | "e" | "se" | "s" | "sw" | "w" | "nw" | "c" | "_" */>?
        get() = definedExternally
        set(value) = definedExternally
}

external interface NodeRefGroup : ASTBaseParent<NodeRef>

external interface OpParams {
    var number : Number
}
external interface EdgeOperator : ASTBaseNode {
    override var type: String /* "->" | "=>" */
    var params : OpParams?
}

external interface EdgeRHSElement {
    var id: dynamic /* NodeRef | EdgeSubgraph */
        get() = definedExternally
        set(value) = definedExternally
    var edgeop: EdgeOperator

}

external interface Edge : ASTBaseParent<Attribute> {
    var from: dynamic /* NodeRef | NodeRefGroup */
        get() = definedExternally
        set(value) = definedExternally
    var targets: Array<EdgeRHSElement> /* JsTuple<to, EdgeRHSElement, Any, Array<EdgeRHSElement>> */
}

external interface Node : ASTBaseParent<Attribute> {
    var id: Literal__0
}


external interface Subgraph : ASTBaseParent<dynamic /* Attribute | Attributes | Edge | Node | Subgraph | Comment */> {
    var id: Literal__0?
        get() = definedExternally
        set(value) = definedExternally

    var specialType: String?
        get() = definedExternally
        set(value) = definedExternally

}

external interface ParseOption<T : Any> {
    var rule: T?
        get() = definedExternally
        set(value) = definedExternally
}

external interface ParseOption__0 : ParseOption<Any>

external fun parse(ocdot: String): OcDot

external fun parse(ocdot: String, options: ParseOption<Any>): dynamic /* OcDot */

external fun parse(
    ocdot: String,
    options: ParseOption__0,
): dynamic /* OcDot | OcNet | Attribute | Attributes | Edge | Node | Subgraph | Comment | Array<dynamic /* Attribute | Attributes | Edge | Node | Subgraph | Comment */> */

external interface StringifyOption {
    var indentSize: Number?
        get() = definedExternally
        set(value) = definedExternally
}

external open class Compiler(__0: StringifyOption = definedExternally) {
    open var indentSize: Number
    open fun indent(line: String): String
    open fun pad(pad: String): (l: String) -> String
    open fun printAttribute(ast: Attribute): String
    open fun printAttributes(ast: Attributes): String
    open fun printComment(ast: Comment): String
    open fun printOcDot(ast: OcDot): String
    open fun printEdge(ast: Edge): String
    open fun printEdgeRHSElement(edgeRHSElement: EdgeRHSElement): String
    open fun printNode(ast: Node): String
    open fun printNodeRef(ast: NodeRef): String
    open fun printNodeRefGroup(ast: NodeRefGroup): String
    open fun printOcNet(ast: OcNet): String
    open fun printSubgraph(ast: Subgraph): String
    open fun printLiteral(ast: Literal__0): String

    open var isAstNode: Any
    open fun stringify(ast: Attribute): String
    open fun stringify(ast: Attributes): String
    open fun stringify(ast: Comment): String
    open fun stringify(ast: OcDot): String
    open fun stringify(ast: Edge): String
    open fun stringify(ast: OcNet): String
    open fun stringify(ast: Literal__0): String
    open fun stringify(ast: Node): String
    open fun stringify(ast: NodeRef): String
    open fun stringify(ast: NodeRefGroup): String
    open fun stringify(ast: Subgraph): String
    open fun stringify(ast: EdgeRHSElement): String
}

external fun stringify(ast: Attribute, options: StringifyOption = definedExternally): String

external fun stringify(ast: Attribute): String

external fun stringify(ast: Attributes, options: StringifyOption = definedExternally): String

external fun stringify(ast: Attributes): String

external fun stringify(ast: Comment, options: StringifyOption = definedExternally): String

external fun stringify(ast: Comment): String

external fun stringify(ast: OcDot, options: StringifyOption = definedExternally): String

external fun stringify(ast: OcDot): String

external fun stringify(ast: Edge, options: StringifyOption = definedExternally): String

external fun stringify(ast: Edge): String

external fun stringify(ast: OcNet, options: StringifyOption = definedExternally): String

external fun stringify(ast: OcNet): String

external fun stringify(ast: Literal__0, options: StringifyOption = definedExternally): String

external fun stringify(ast: Literal__0): String

external fun stringify(ast: Node, options: StringifyOption = definedExternally): String

external fun stringify(ast: Node): String

external fun stringify(ast: NodeRef, options: StringifyOption = definedExternally): String

external fun stringify(ast: NodeRef): String

external fun stringify(ast: NodeRefGroup, options: StringifyOption = definedExternally): String

external fun stringify(ast: NodeRefGroup): String

external fun stringify(ast: Subgraph, options: StringifyOption = definedExternally): String

external fun stringify(ast: Subgraph): String

File: ocdot-parser/src/ast.ts
import { Compass, Int } from 'ts-graphviz';
import { parse as _parse, FileRange as _FileRange } from './ocdot.peggy';
import { prependIndent } from './exts';

/**
 * The `AST` module provides the ability to handle the AST as a result of parsing the ocdot language
 * for lower level operations.
 *
 * @alpha
 */
export namespace AST {
  export type FileRange = _FileRange;
  type ValueOf<T> = T[keyof T];

  /**
   * DOT object types.
   */
  export const Types = Object.freeze({
    OcDot: 'ocdot',
    Comment: 'comment',
    Ocnet: 'ocnet',
    Attribute: 'attribute',
    Attributes: 'attributes',
    Edge: 'edge',
    Node: 'node',
    NodeRef: 'node_ref',
    // NodeRefGroup: 'node_ref_group',
    Subgraph: 'subgraph',
    EdgeSubgraph: 'edge_subgraph',
    Literal: 'literal',
    ClusterStatements: 'cluster_statements',
    TypeDefinitions: 'type_definitions'
  } as const);
  // 'ocdot' | 'comment' etc
  export type Types = ValueOf<typeof Types>;

  export const SubgraphSpecialTypes = Object.freeze({
    Places: 'places',
    Transitions: 'transitions',
  } as const)
  export type SubgraphSpecialTypes = ValueOf<typeof SubgraphSpecialTypes>

  export const OpTypes = Object.freeze({
    Normal: '->',
    Variable: '=>'
  } as const)
  export type OpTypes = ValueOf<typeof OpTypes>

  export const OpParamsTypes = Object.freeze({
    Expression : "expression",
    Number : "number"
  })
  export type OpParamsTypes = ValueOf<typeof OpParamsTypes>

  export function isASTBaseNode(value: unknown): value is ASTBaseNode {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return typeof value === 'object' && value !== null && typeof (value as any).type === 'string';
  }


  /**
   * AST node.
   */
  export interface ASTBaseNode {
    /**
     * Every leaf interface that extends ASTBaseNode
     * must specify a type property.
     */
    type: string;
    location: FileRange;
  }

  export interface ASTBaseParent<STMT extends ASTBaseNode = ASTBaseNode> extends ASTBaseNode {
    body: STMT[];
  }

  export interface Literal<T extends string = string> extends ASTBaseNode {
    type: typeof Types.Literal;
    value: T;
    quoted: boolean | 'html';
  }

  export interface OcDot extends ASTBaseParent<OcDotStatement> {
    type: typeof Types.OcDot;
  }

  /**
   * Graph AST object.
   */
  export interface OcNet extends ASTBaseParent<ClusterStatement> {
    type: typeof Types.Ocnet;
    id?: Literal;
  }

  export interface KeyValue {
    key: Literal;
    value: Literal;
  }

  /**
   * Attribute AST object.
   */
  export interface Attribute extends ASTBaseNode, KeyValue {
    type: typeof Types.Attribute;
  }

  /**
   * Comment AST object.
   */
  export interface Comment extends ASTBaseNode {
    type: typeof Types.Comment;
    kind: Comment.Kind;
    value: string;
  }
  export namespace Comment {
    export const Kind = Object.freeze({
      Block: 'block',
      Slash: 'slash',
      Macro: 'macro',
    } as const);
    export type Kind = ValueOf<typeof Kind>;
  }

  /** Attributes AST object. */
  export interface Attributes extends ASTBaseParent<Attribute | Comment> {
    type: typeof Types.Attributes;
    kind: Attributes.Kind;
  }
  export namespace Attributes {
    export const Kind = Object.freeze({
      Ocnet: Types.Ocnet,
      Edge: Types.Edge,
      Node: Types.Node,
    } as const);
    export type Kind = ValueOf<typeof Kind>;
  }

  /** NodeRef AST object. */
  export interface NodeRef extends ASTBaseNode {
    type: typeof Types.NodeRef;
    id: Literal;
    port?: Literal;
    compass?: Literal<Compass>;
  }

  /** NodeRefGroup AST object. */
  // export interface NodeRefGroup extends ASTBaseParent<NodeRef> {
  //   type: typeof Types.NodeRefGroup;
  // }

  export interface EdgeSubgraph extends ASTBaseParent<ClusterStatement> {
    type: typeof Types.EdgeSubgraph;
    id?: Literal;
  }

  export type EdgeTarget = NodeRef | EdgeSubgraph;

  export interface NumberLiteral extends ASTBaseNode {
    value: number,
  }

  export interface Variable { 
    variable : string
  }

  export interface ExpressionOp {
    op : "*" | "+" | "-" | "/",
    target : Expression
  }

  export interface Expression { 
    head : Expression | number | Variable
    tail : ExpressionOp[]
  }

  export interface RootExpression extends EdgeOpParams, Expression { 
    type : typeof OpParamsTypes.Expression
  }

  export interface EdgeOpParams extends ASTBaseNode {
    type : OpParamsTypes,
  }

  export interface OpParamsNumber extends EdgeOpParams {
    type : typeof OpParamsTypes.Number
    value : number
  }

  export interface EdgeOperator extends ASTBaseNode {
    type: "->" | "=>",
    params?: EdgeOpParams
  }

  export interface EdgeRHSElement {
    id: EdgeTarget,
    edgeop: EdgeOperator
  }


  /** Edge AST object. */
  export interface Edge extends ASTBaseParent<Attribute> {
    type: typeof Types.Edge;
    from: EdgeTarget,
    targets: [to: EdgeRHSElement, ...rest: EdgeRHSElement[]];
  }

  /** Node AST object. */
  export interface Node extends ASTBaseParent<Attribute> {
    type: typeof Types.Node;
    id: Literal;
  }

  /** Subgraph AST object. */
  export interface Subgraph extends ASTBaseParent<ClusterStatement> {
    type: typeof Types.Subgraph;
    id?: Literal;
    specialType?: SubgraphSpecialTypes
  }

  export type TypeDefinition = Node | Comment;
  export type OcDotStatement = OcNet | Comment;
  export type ClusterStatement = Attribute | Attributes | Edge | Node | Subgraph | Comment;

  export type ASTNode =
    | Attribute
    | Attributes
    | Comment
    | OcDot
    | Edge
    | OcNet
    | Literal
    | Node
    | NodeRef
    | EdgeSubgraph
    | Subgraph;

  export type Rule =
    | typeof Types.OcDot
    | typeof Types.Ocnet
    | typeof Types.Node
    | typeof Types.Edge
    | typeof Types.Attributes
    | typeof Types.Subgraph
    | typeof Types.Attribute
    | typeof Types.ClusterStatements;

  /**
   * Option interface for {@link parse} function.
   */
  export interface ParseOption<T extends Rule = Rule> {
    rule?: T;
  }

  /**
   * The basic usage is the same as the `parse` function,
   * except that it returns the ocdot language
   *
   * ```ts
   * import { AST } from '@ts-graphviz/parser';
   *
   * const ast = parse(`
   *   digraph example {
   *     node1 [
   *       label = "My Node",
   *     ]
   *   }
   * `);
   *
   * console.log(ast);
   * // {
   * //   type: 'ocdot',
   * //   body: [
   * //     {
   * //       type: 'graph',
   * //       id: {
   * //         type: 'literal',
   * //         value: 'example',
   * //         quoted: false,
   * //         location: {
   * //           start: { offset: 11, line: 2, column: 11 },
   * //           end: { offset: 18, line: 2, column: 18 }
   * //         }
   * //       },
   * //       directed: true,
   * //       strict: false,
   * //       body: [
   * //         {
   * //           type: 'node',
   * //           id: {
   * //             type: 'literal',
   * //             value: 'node1',
   * //             quoted: false,
   * //             location: {
   * //               start: { offset: 25, line: 3, column: 5 },
   * //               end: { offset: 30, line: 3, column: 10 }
   * //             }
   * //           },
   * //           body: [
   * //             {
   * //               type: 'attribute',
   * //               key: {
   * //                 type: 'literal',
   * //                 value: 'label',
   * //                 quoted: false,
   * //                 location: {
   * //                   start: { offset: 39, line: 4, column: 7 },
   * //                   end: { offset: 44, line: 4, column: 12 }
   * //                 }
   * //               },
   * //               value: {
   * //                 type: 'literal',
   * //                 value: 'My Node',
   * //                 quoted: true,
   * //                 location: {
   * //                   start: { offset: 47, line: 4, column: 15 },
   * //                   end: { offset: 56, line: 4, column: 24 }
   * //                 }
   * //               },
   * //               location: {
   * //                 start: { offset: 39, line: 4, column: 7 },
   * //                 end: { offset: 57, line: 4, column: 25 }
   * //               }
   * //             }
   * //           ],
   * //           location: {
   * //             start: { offset: 25, line: 3, column: 5 },
   * //             end: { offset: 63, line: 5, column: 6 }
   * //           }
   * //         }
   * //       ],
   * //       location: {
   * //         start: { offset: 3, line: 2, column: 3 },
   * //         end: { offset: 67, line: 6, column: 4 }
   * //       }
   * //     }
   * //   ],
   * //   location: {
   * //     start: { offset: 3, line: 2, column: 3 },
   * //     end: { offset: 68, line: 7, column: 1 }
   * //   }
   * // }
   * ```
   *
   * @param ocdot string in the ocdot language to be parsed.
   * @param options.rule Object type of ocdot string.
   * This can be "graph", "subgraph", "node", "edge",
   * "attributes", "attribute", "cluster_statements".
   *
   * @example
   * const ast = parse('test [ style=filled; ];', { rule: 'node' });
   *
   * console.log(ast);
   * // {
   * //   type: 'node',
   * //   id: {
   * //     type: 'literal',
   * //     value: 'test',
   * //     quoted: false,
   * //     location: {
   * //       start: { offset: 0, line: 1, column: 1 },
   * //       end: { offset: 4, line: 1, column: 5 }
   * //     }
   * //   },
   * //   body: [
   * //     {
   * //       type: 'attribute',
   * //       key: {
   * //         type: 'literal',
   * //         value: 'style',
   * //         quoted: false,
   * //         location: {
   * //           start: { offset: 7, line: 1, column: 8 },
   * //           end: { offset: 12, line: 1, column: 13 }
   * //         }
   * //       },
   * //       value: {
   * //         type: 'literal',
   * //         value: 'filled',
   * //         quoted: false,
   * //         location: {
   * //           start: { offset: 13, line: 1, column: 14 },
   * //           end: { offset: 19, line: 1, column: 20 }
   * //         }
   * //       },
   * //       location: {
   * //         start: { offset: 7, line: 1, column: 8 },
   * //         end: { offset: 20, line: 1, column: 21 }
   * //       }
   * //     }
   * //   ],
   * //   location: {
   * //     start: { offset: 0, line: 1, column: 1 },
   * //     end: { offset: 23, line: 1, column: 24 }
   * //   }
   * // }
   *
   * @returns The AST object of the parse result is returned.
   *
   * @throws {SyntaxError}
   */
  export function parse(ocdot: string): OcDot;
  export function parse(ocdot: string, options: ParseOption<typeof Types.Edge>): Edge;
  export function parse(ocdot: string, options: ParseOption<typeof Types.Node>): Node;
  export function parse(ocdot: string, options: ParseOption<typeof Types.Ocnet>): OcNet;
  export function parse(ocdot: string, options: ParseOption<typeof Types.Attribute>): Attribute;
  export function parse(ocdot: string, options: ParseOption<typeof Types.Attributes>): Attributes;
  export function parse(ocdot: string, options: ParseOption<typeof Types.Subgraph>): Subgraph;
  export function parse(ocdot: string, options: ParseOption<typeof Types.ClusterStatements>): ClusterStatement[];
  export function parse(ocdot: string, options: ParseOption<typeof Types.OcDot>): OcDot;
  export function parse(ocdot: string, options: ParseOption): OcDot | OcNet | ClusterStatement | ClusterStatement[];
  export function parse(ocdot: string, { rule }: ParseOption = {}): OcDot | OcNet | ClusterStatement | ClusterStatement[] {
    return _parse(ocdot, {
      startRule: rule,
    });
  }

  export function parseExpression(expression : string) : RootExpression {
    return _parse(expression, {
      startRule: "RootExpression"
    })
  }

  export interface StringifyOption {
    indentSize?: number;
  }

  
  export function isEdgeOpParamsExpression(edgeOpParams : EdgeOpParams): edgeOpParams is RootExpression { 
    return edgeOpParams.type == OpParamsTypes.Expression
  }

  export function isEdgeOpParamsNumber(edgeOpParams : EdgeOpParams) : edgeOpParams is OpParamsNumber {
    return edgeOpParams.type == OpParamsTypes.Number
  }

  export function isExpression(item : any) : item is Expression {
    if (typeof item !== "object") {
      return false
    }
    return "head" in item
  }

  export function isVariable(object : any) : object is Variable { 
    if (typeof object !== "object") {
      return false
    }
    return "variable" in object
  }

  export class Compiler {
    protected indentSize: number;
    constructor({ indentSize = 0 }: StringifyOption = {}) {
      this.indentSize = indentSize;
    }

    protected indent(line: string): string {
      return '  '.repeat(this.indentSize) + line;
    }

    protected indentForSize(size : number) : string {
      return '  '.repeat(size);
    }

    protected buildIndent(): string {
      return this.indentForSize(this.indentSize)
    }

    protected withIndentIncrease(multiline: string): string {
      // this.increaseIndent()
      const saved = this.indentSize
      this.indentSize = 2
      const value = prependIndent(multiline, this.buildIndent())
      // this.decreaseIndent()
      this.indentSize = saved;
      return value
    }

    protected pad(pad: string): (l: string) => string {
      return (l: string) => pad + l;
    }

    protected printAttribute(ast: AST.Attribute): string {
      return `${this.stringify(ast.key)} = ${this.stringify(ast.value)};`;
    }

    protected increaseIndent() {
      this.indentSize += 2;
    }

    protected decreaseIndent() {
      this.indentSize -= 2;
    }

    protected printAttributes(ast: AST.Attributes): string {
      const value = ast.body.length === 0
        ? `${ast.kind};`
        : `${ast.kind} [\n${this.withIndentIncrease(ast.body.map(this.stringify, this).join('\n'))}\n];`;
      return value;
    }

    protected printComment(ast: AST.Comment): string {
      switch (ast.kind) {
        case AST.Comment.Kind.Block:
          return '/**\n' + ast.value.split('\n').map(this.pad(' * ')).join('\n') + '\n */';
        case AST.Comment.Kind.Slash:
          return ast.value.split('\n').map(this.pad('// ')).join('\n');
        case AST.Comment.Kind.Macro:
          return ast.value.split('\n').map(this.pad('# ')).join('\n');
      }
    }
    protected printOcDot(ast: AST.OcDot): string {
      return ast.body.map(this.stringify.bind(this)).join('\n');
    }

    protected printEdge(ast: AST.Edge): string {
      const edgeOp = ast
      const from = this.stringify(ast.from)
      const targets = ast.targets.map(this.stringify.bind(this))/* .join(this.directed ? ' -> ' : ' -- '); */
      const allEdgeTargets = [from, ...targets].join(' ')

      if (ast.body.length === 0) {
        return allEdgeTargets + ";"
      }

      return `${allEdgeTargets} [\n${this.withIndentIncrease(ast.body.map(this.stringify, this)
        .join('\n'))
        }\n]`
    }

    protected stringifyExpressionElement(head : Expression | number | Variable) : string {
      if (isExpression(head)) {
        return this.stringifyExpression(head);
      } else if (isVariable(head)) {
        return head.variable
      } else {
        return head.toString()
      }
    }

    protected stringifyExpressionOp(expressionOp : ExpressionOp) : string { 
      let op = expressionOp.op
      let expression = this.stringifyExpressionElement(expressionOp.target)
      return `${op} ${expression}`
    }

    public stringifyExpression(expression : Expression) : string {
      let arr = [ ]
      if (expression.tail.length == 0 && !isExpression(expression.head)) {
        return this.stringifyExpressionElement(expression.head);
      }
      arr.push(this.stringifyExpressionElement(expression.head))
      for (let i = 0; i < expression.tail.length; i++) {
        arr.push(this.stringifyExpressionOp(expression.tail[i]))
      }
      return "(" + arr.join(' ') + ")";
    }

    protected printEdgeOpParams(edgeOpParams : EdgeOpParams) : string {
      if (isEdgeOpParamsExpression(edgeOpParams)) {
        return this.stringifyExpression(edgeOpParams)
      } else if (isEdgeOpParamsNumber(edgeOpParams))  {
        return edgeOpParams.value.toString()
      } else {
        return "UNKNOWN_EDGE_OP_PARAM"
      }
    }

    protected printEdgeRHSElement(edgeRHSElement: EdgeRHSElement): string {
      const edgeOp = edgeRHSElement.edgeop.type
      const multiplicity = edgeRHSElement.edgeop.params
        ? this.printEdgeOpParams(edgeRHSElement.edgeop.params)
        : ""
      return `${multiplicity}${edgeOp} ${this.stringify(edgeRHSElement.id)}`
    }

    protected printNode(ast: AST.Node): string {
      if (ast.body.length === 0) {
        return this.stringify(ast.id)
      }
      return `${this.stringify(ast.id)} [\n${this.withIndentIncrease(ast.body.map(this.stringify, this)
        .join('\n'))
        }\n];`
    }

    protected printNodeRef(ast: AST.NodeRef): string {
      return [
        this.stringify(ast.id),
        ast.port ? this.stringify(ast.port)
          : null,
        ast.compass ? this.stringify(ast.compass) : null,
      ]
        .filter((v) => v !== null)
        .join(':');
    }

    protected printEdgeSubgraphName(ast: AST.EdgeSubgraph): (string | null)[] {
      if (ast.id == null) {
        return []
      } else {
        return ['subgraph', this.stringify(ast.id)]
      }
    }

    protected printEdgeSubgraph(ast: AST.EdgeSubgraph): string {
      if (ast.body.length === 0) {
        return "{}"
      }
      const body = `{\n${this.withIndentIncrease(ast.body.map(this.stringify, this).join('\n'))
        }\n}`

      return [
        ...this.printEdgeSubgraphName(ast),
        body
      ]
        .filter((v) => v !== null)
        .join(' ');
    }

    protected printOcNet(ast: AST.OcNet): string {
      var body = "ocnet {"
      if (ast.body.length === 0) {
        body += "}"
        return body
      }
      body += "\n"
      body += this.withIndentIncrease(ast.body
        .map(this.stringify.bind(this))
        .join('\n'))
      body += "\n}"

      return [
        // ast.strict ? 'strict' : null,
        // ast.directed ? 'digraph' : 'graph',
        ast.id ? this.stringify(ast.id) : null,
        body,
      ]
        .filter((v) => v !== null)
        .join(' ');
    }

    protected checkSubgraphKeyword(ast: AST.Subgraph): boolean {
      return ast.specialType != null
    }

    protected printSubgraphName(ast: AST.Subgraph): (string | null)[] {
      if (this.checkSubgraphKeyword(ast)) {
        return [ast.specialType ?? "", ast.id ? this.stringify(ast.id) : null]
      } else {
        return ['subgraph', ast.id ? this.stringify(ast.id) : null]
      }
    }

    protected printSubgraph(ast: AST.Subgraph): string {
      var body = "{"
      if (ast.body.length === 0) {
        body += "}"
        return body
      }
      body += "\n"
      body += this.withIndentIncrease(
        ast.body.map(this.stringify.bind(this)).join('\n')
      )
      body += "\n}"

      return [
        ...this.printSubgraphName(ast),
        body
      ]
        .filter((v) => v !== null)
        .join(' ');
    }

    protected printLiteral(ast: AST.Literal): string {
      switch (ast.quoted) {
        case true:
          return `"${ast.value}"`;
        case false:
          return ast.value;
        case 'html':
          return `<${ast.value}>`;
      }
    }


    protected isAstNode(object: any): object is AST.ASTNode {
      if (typeof object !== "object") {
        return false
      }
      return 'type' in object;
    }

    public stringify(ast: AST.ASTNode | EdgeRHSElement): string {
      if (this.isAstNode(ast)) {
        switch (ast.type) {
          case AST.Types.Attribute:
            return this.printAttribute(ast);
          case AST.Types.Attributes:
            return this.printAttributes(ast);
          case AST.Types.Comment:
            return this.printComment(ast);
          case AST.Types.OcDot:
            return this.printOcDot(ast);
          case AST.Types.Edge:
            return this.printEdge(ast);
          case AST.Types.Node:
            return this.printNode(ast);
          case AST.Types.NodeRef:
            return this.printNodeRef(ast);
          case AST.Types.EdgeSubgraph:
            return this.printEdgeSubgraph(ast);
          case AST.Types.Ocnet:
            // this.directed = ast.directed;
            return this.printOcNet(ast);
          case AST.Types.Subgraph:
            return this.printSubgraph(ast);
          case AST.Types.Literal:
            return this.printLiteral(ast);
        }
      } else {
        return this.printEdgeRHSElement(ast)
      }
    }
  }

  /**
   * Stringify Graphviz AST Node.
   *
   * @param ast Graphviz AST node.
   * @returns DOT language string.
   */
  export function stringify(ast: AST.ASTNode, options?: StringifyOption): string {
    return new Compiler(options).stringify(ast);
  }
}
